extensions/amp-inline-gallery/1.0/amp-inline-gallery-thumbnails.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../amp-base-carousel/1.0/base-carousel.jss';
+import {CarouselContextProp} from '../../amp-base-carousel/1.0/carousel-props';
+import {Layout} from '../../../src/layout';
+import {PreactBaseElement} from '../../../src/preact/base-element';
+import {Thumbnails} from './thumbnails';
+import {
+  closestAncestorElementBySelector,
+  matches,
+  scopedQuerySelector,
+} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {toArray} from '../../../src/types';
+
+/** @const {string} */
+export const TAG = 'amp-inline-gallery-thumbnails';
+
+/**
+ * The selector for the main carousel (i.e. not the one for the thumbnails).
+ */
+const CAROUSEL_SELECTOR =
+  '> amp-base-carousel, :not(amp-inline-gallery-thumbnails) > amp-base-carousel';
+
+const GALLERY_SELECTOR = 'amp-inline-gallery';
+
+/** @extends {PreactBaseElement<BaseCarouselDef.CarouselApi>} */
+export class AmpInlineGalleryThumbnails extends PreactBaseElement {
+  /** @override */
+  init() {
+    const inlineGallery = dev().assertElement(
+      closestAncestorElementBySelector(this.element, GALLERY_SELECTOR)
+    );
+    const carousel = scopedQuerySelector(inlineGallery, CAROUSEL_SELECTOR);
+    const slides = toArray(carousel.children);
+    slides.forEach((slide) => {
+      this.element.appendChild(createThumbnailForElement(slide));
```

Comment:
There might be even more issues here. E.g. a slide's image might change its `src` and it'd be hard to re-render them correctly here. It would have been great if we could simply pass these elements as Preact components to be rendered by Preact itself. WDYT? E.g. we can simply pass `(<img src={...}>)` and let Preact figure it all out. We lose the `slide.cloneNode()` below, but maybe it's a good thing?